### PR TITLE
Core: exclude NaN from upper/lower bound of floating columns in Parquet/ORC

### PR DESCRIPTION
**Pull request by yyanyy**
- This PR implements the behavior described in #348, to exclude NaN from upper/lower bound of double/float type columns in parquet/orc. 
- Support for Avro metrics can be found in #1963, and has some overlap with this PR (especially on `FieldMetrics` related classes)

### CODE REVIEW

File: core/src/main/java/org/apache/iceberg/FieldMetrics.java [hunk 1]
```diff
@@ -20,25 +20,23 @@
 package org.apache.iceberg;
 
 
-import java.nio.ByteBuffer;
-
 /**
  * Iceberg internally tracked field level metrics.
  */
-public class FieldMetrics {
+public class FieldMetrics<T> {
```
Comment (yyanyy) [inline core/src/main/java/org/apache/iceberg/FieldMetrics.java:26]: Change in this class overlaps with what #1963 contains. For now this class is not directly used/created by Parquet/ORC, and will be for Avro it will, thus should be safe to change. Same goes for `FloatFieldMetrics`.

File: data/src/test/java/org/apache/iceberg/TestMergingMetrics.java [hunk 2]
```diff
@@ -105,74 +117,152 @@ public TestMergingMetrics(FileFormat fileFormat) {
...
+
+  private void assertNaNCountMatch(Long expected, Map<Integer, Long> nanValueCount, Types.NestedField field) {
+    Assert.assertEquals(
+        String.format("NaN count for field %s does not match expected", field.name()),
+        expected, nanValueCount.get(FIELDS_WITH_NAN_COUNT_TO_ID.get(field)));
+  }
+
+  private void assertBoundValueMatch(Number expected, Map<Integer, ByteBuffer> boundMap, Types.NestedField field) {
+    if (field.type().isNestedType() && fileFormat == FileFormat.ORC) {
+      // we don't update floating column bounds values within ORC nested columns
```
Comment (yyanyy) [inline data/src/test/java/org/apache/iceberg/TestMergingMetrics.java:186]: We actually could make it work for Orc but I'm not sure if it's worth it given metrics in nested types are not recorded/used normally anyway. The reason for the different behavior in Parquet/Orc is the following:

Parquet tracks column stats per block, and when computing metrics we need to update min/max when iterating through all blocks; to avoid doing so for floating point columns that we track min/max ourselves now, we update min/max for these columns after all blocks have been processed. 

However Orc tracks stats per column so we could directly inject the new bound value when we evaluates every column stats. And since Orc doesn't track nested column stats, we never look at columns within nested fields.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 3]
```diff
@@ -85,6 +89,7 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
     return footerMetrics(metadata, fieldMetrics, metricsConfig, null);
   }
 
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
```
Comment (yyanyy) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:92]: Decided to not refactor this method (at least not in this PR) since in this method we are essentially updating all column stats, and by moving logic directly out of the method we need to pass in those 6 stats map with some extra info, which could easily make the helper method exceeding 10 arguments.

File: data/src/test/java/org/apache/iceberg/TestMergingMetrics.java [hunk 4]
```diff
@@ -37,6 +48,7 @@
 
 import static org.apache.iceberg.types.Types.NestedField.optional;
 import static org.apache.iceberg.types.Types.NestedField.required;
+import static org.junit.Assert.assertEquals;
```
Comment (rdblue) [inline data/src/test/java/org/apache/iceberg/TestMergingMetrics.java:51]: Nit: the community prefers to avoid static method imports. There are places (usually my fault) that still import optional and required like the lines above, but we tend to discourage doing this in new pull requests. Could you remove this and just import `Assert`?

File: data/src/test/java/org/apache/iceberg/TestMergingMetrics.java (hunk 2 above)
Comment (rdblue) [reply data/src/test/java/org/apache/iceberg/TestMergingMetrics.java:186]: I think it's okay to leave this as-is. I actually don't think that we really need these stats for Parquet either, but it's probably better to have them.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 5]
```diff
@@ -125,7 +134,9 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
         } else if (!stats.isEmpty()) {
           increment(nullValueCounts, fieldId, stats.getNumNulls());
 
-          if (metricsMode != MetricsModes.Counts.get()) {
+          // since Parquet includes NaN for upper/lower bounds of floating point column that we don't want,
+          // we have tracked metrics for such columns ourselves and thus do not need to rely on Parquet's stats.
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:138]: You might want to rephrase this not to assume the reason why `fieldMetricsMap` contains metrics. If there are metrics gathered by Iceberg for a column, then we should use those instead of the ones from Parquet. The NaN handling for float and double columns is one case of that. But there may be others in the future.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 6]
```diff
@@ -147,9 +158,40 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
...
+          lowerBounds.remove(fieldId);
+          upperBounds.remove(fieldId);
+        } else if (metrics.upperBound() instanceof Float) {
+          lowerBounds.put(fieldId, Literal.of((Float) metrics.lowerBound()));
+          upperBounds.put(fieldId, Literal.of((Float) metrics.upperBound()));
+        } else if (metrics.upperBound() instanceof Double) {
+          lowerBounds.put(fieldId, Literal.of((Double) metrics.lowerBound()));
+          upperBounds.put(fieldId, Literal.of((Double) metrics.upperBound()));
+        } else {
+          throw new IllegalStateException("[BUG] Only Float/Double column metrics should be collected by Iceberg");
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:191]: Minor: I would probably rephrase this as "Expected only float or double column metrics" and throw an `UnsupportedOperationException`. I think that is (marginally) more clear because it isn't quite right to say that only float metrics should be collected. The code assumed that, but there isn't a reason for that to be the case other than we didn't implement the other types.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 7]
```diff
@@ -147,9 +158,40 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
...
+
     return new Metrics(rowCount, columnSizes, valueCounts, nullValueCounts,
-        MetricsUtil.createNanValueCounts(fieldMetrics, metricsConfig, fileSchema),
-        toBufferMap(fileSchema, lowerBounds), toBufferMap(fileSchema, upperBounds));
+        MetricsUtil.createNanValueCounts(fieldMetricsMap.values().stream(), metricsConfig, fileSchema),
+        toBufferMap(fileSchema, lowerBounds),
+        toBufferMap(fileSchema, upperBounds));
+  }
+
+  private static void updateFloatingColumnsBounds(
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:170]: Minor: You might consider renaming this so that it doesn't assume that Iceberg is maintaining bounds only for floating point columns. Maybe `updateFromFieldMetrics`?

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 8]
```diff
@@ -99,6 +104,10 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
     MessageType parquetTypeWithIds = getParquetTypeWithIds(metadata, nameMapping);
     Schema fileSchema = ParquetSchemaUtil.convertAndPrune(parquetTypeWithIds);
 
+    Map<Integer, FieldMetrics> fieldMetricsMap = Optional.ofNullable(fieldMetrics)
+        .map(stream -> stream.collect(Collectors.toMap(FieldMetrics::id, Function.identity())))
+        .orElseGet(HashMap::new);
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:109]: Why not `.orElse(ImmutableMap.of())`? Seems strange to create a new map if field metrics are missing. And won't `stream.collect(Collectors.toMap(...))` return an empty map if the stream is empty? Does this need to handle a null stream or can it just assume that it will be non-null?

File: core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java [hunk 9]
```diff
@@ -19,25 +19,17 @@
...
-
-  /**
-   * Constructor for creating a FieldMetrics with only NaN counter.
-   * @param id field id being tracked by the writer
-   * @param nanValueCount number of NaN values, will only be non-0 for double or float field.
-   */
-  public FloatFieldMetrics(int id,
-                           long nanValueCount) {
-    super(id, 0L, 0L, nanValueCount, null, null);
+public class FloatFieldMetrics extends FieldMetrics<Number> {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java:29]: Since this is a public class, I think it would make more sense to split it into an implementation for float and an implementation for double and have each implementation correctly set the type that is tracked. That avoids needing to test the type of upper bound or lower bound because you know that both are going to be the same. That's an assumption that the code below makes, but it isn't necessarily the case.

File: core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java [hunk 10]
```diff
@@ -50,13 +42,64 @@ public long nullValueCount() {
...
+          this.upperBound = value;
+        }
+      }
+    }
   }
 
-  @Override
-  public ByteBuffer upperBound() {
-    throw new IllegalStateException("Shouldn't access this method, as this metric is tracked in file statistics. ");
+  public static class DoubleFieldMetricsContext extends AbstractFloatFieldMetricsContext<Double> {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java:65]: This class is suspiciously similar to the builder pattern, so I would probably just convert it to be a builder for `FieldMetrics` instead:

* Rename `DoubleFieldMetricsContext` to `DoubleFieldMetrics.Builder` and create it with `builderFor(int id)`
* Rename `updateMetricsContext` to something shorter, like `addValue`
* Rename `buildMetrics` to just `build`

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java (hunk 8 above)
Comment (yyanyy) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java]: I was trying to be defensive in handling null stream here which is why I had this Optional wrapper and `orElse`, but you are right the current implementation wouldn't produce a null stream. I'll add a precondition check for null stream and then directly use the stream to create the map.

File: core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java (hunk 9 above)
Comment (yyanyy) [reply core/src/main/java/org/apache/iceberg/FloatFieldMetrics.java]: Sounds good! I was hesitant to do that since I think there are too many duplicated code, I guess I was trying too hard to eliminate duplications...

File: core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+ */
+public class DoubleFieldMetrics extends FieldMetrics<Double> {
+
+  private DoubleFieldMetrics(int id, long nanValueCount, Double lowerBound, Double upperBound) {
+    super(id, 0L, 0L, nanValueCount, lowerBound, upperBound);
+  }
+
+  @Override
+  public long valueCount() {
+    throw new IllegalStateException("Shouldn't access this method, as this metric is tracked in file statistics. ");
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java:37]: We typically prefer `UnsupportedOperationException` for cases like this rather than `IllegalStateException`, which is described in javadoc as an exception that "Signals that a method has been invoked at an illegal or inappropriate time". While you could argue that any time is an "inappropriate time", I think unsupported describes the situation better.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java:37]: Thinking about this a bit more, would it be a bad thing to collect the stats so that future uses don't need to go back and modify this class? Seems like there isn't much benefit to skipping the implementation for just these two methods.

File: core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+    public Builder(int id) {
+      this.id = id;
+    }
+
+    public void addValue(double value) {
+      if (Double.isNaN(value)) {
+        this.nanValueCount++;
+      } else {
+        if (lowerBound == null || Double.compare(value, lowerBound) < 0) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java:59]: I mentioned this on the Avro PR, but I think it would be better to make `lowerBound` and `upperBound` primitives and get rid of the null checks here. Then you can use `valueCount > 0` to determine whether to return the bounds or null.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 13]
```diff
@@ -99,6 +105,9 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
     MessageType parquetTypeWithIds = getParquetTypeWithIds(metadata, nameMapping);
     Schema fileSchema = ParquetSchemaUtil.convertAndPrune(parquetTypeWithIds);
 
+    Map<Integer, FieldMetrics> fieldMetricsMap = fieldMetrics.collect(
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:108]: `FieldMetrics` is parameterized, right? I think this should be `FieldMetrics<?>`.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 14]
```diff
@@ -147,9 +158,39 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
...
+      FieldMetrics metrics = entry.getValue();
+      MetricsMode metricsMode = MetricsUtil.metricsMode(schema, metricsConfig, fieldId);
+
+      // only check for MetricsModes.None, since we don't truncate float/double values.
+      if (metricsMode != MetricsModes.None.get()) {
+        if (metrics.upperBound() == null) {
+          // upper and lower bounds will both null or neither
+          lowerBounds.remove(fieldId);
+          upperBounds.remove(fieldId);
+        } else if (metrics.upperBound() instanceof Float) {
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:183]: I think this should be `metrics instanceof FloatFieldMetrics` instead of checking the bound.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java [hunk 15]
```diff
@@ -147,9 +158,39 @@ public static Metrics footerMetrics(ParquetMetadata metadata, Stream<FieldMetric
...
+      Map<Integer, FieldMetrics> idToFieldMetricsMap, MetricsConfig metricsConfig, Schema schema,
+      Map<Integer, Literal<?>> lowerBounds, Map<Integer, Literal<?>> upperBounds) {
+    idToFieldMetricsMap.entrySet().forEach(entry -> {
+      int fieldId = entry.getKey();
+      FieldMetrics metrics = entry.getValue();
+      MetricsMode metricsMode = MetricsUtil.metricsMode(schema, metricsConfig, fieldId);
+
+      // only check for MetricsModes.None, since we don't truncate float/double values.
+      if (metricsMode != MetricsModes.None.get()) {
+        if (metrics.upperBound() == null) {
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:179]: Can `metrics` expose a better way to detect this than `null`? What about `hasBounds()`?

File: core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java (hunk 11 above)
Comment (yyanyy) [reply core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java:37]: We had to spend some extra effort on nullable writers to make them update null count and value count in order to get the right results for the unimplemented methods, but that's not complicated and will update.

File: core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java (hunk 12 above)
Comment (yyanyy) [reply core/src/main/java/org/apache/iceberg/DoubleFieldMetrics.java:59]: Sorry haven't checked the avro PR yet so didn't see that comment... writers creating this already has null values filtered out so null value count is always 0, but the idea could still work. Will update!

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java (hunk 13 above)
Comment (yyanyy) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:108]: I will modify the ones for Parquet and ORC, and leave the Avro related ones as is, since we are going to modify them in the Avro metrics support PR anyway. Let me know if it works!

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java (hunk 14 above)
Comment (yyanyy) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:183]: Actually due to having to modify the option writer to return correct value count/null value count, this can go different directions:
- we either let the option writer to explicitly check for float/double writer and create the corresponding field metric object type (also have to make float/double field metrics' constructor public)
- or we create the generic field metric type in option writer, and continue to check instanceof on bounds here

For now I decided latter for better flexibility, but don't have strong opinion either way. Please let me know your thoughts!

File: core/src/main/java/org/apache/iceberg/FieldMetrics.java [hunk 16]
```diff
@@ -78,14 +76,21 @@ public long nanValueCount() {
...
-  public ByteBuffer upperBound() {
+  public T upperBound() {
     return upperBound;
   }
+
+  /**
+   * Returns if the metrics has bounds (i.e. there is at least non-null value for this field)
+   */
+  public boolean hasBounds() {
+    return upperBound != null;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/FieldMetrics.java:94]: I think this should be `valueCount - nullValueCount - nanValueCount > 0`. Otherwise, using `DoubleFieldMetrics.Builder` will always produce bounds, even though it could be only NaN values.

Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/FieldMetrics.java:94]: Oh, nevermind, I see that `build` handles that case.

File: data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java [hunk 17]
```diff
@@ -235,24 +235,30 @@ private FloatWriter(int id) {
...
       ((DoubleColumnVector) output).vector[rowId] = data;
-      if (Float.isNaN(data)) {
-        nanCount++;
-      }
+      floatFieldMetricsBuilder.addValue(data);
+    }
+
+    @Override
+    public void nullWrite() {
+      nullValueCount++;
```
Comment (rdblue) [inline data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java:243]: It seems like it may be a good idea to move null tracking into the builder, so you could just run `metricsBuilder.addValue(null)` or `metricsBuilder.addNullValue()`. Then you wouldn't need to rebuild the metrics later.

File: parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java (hunk 14 above)
Comment (rdblue) [reply parquet/src/main/java/org/apache/iceberg/parquet/ParquetUtil.java:182]: Oh, I see. Good point. Let's go with this for now. In the future, we could keep the value class in the metrics, so you could do something like this:

```java
  if (metrics.javaType() instanceof Float) { ... }
```

But that's probably not worth it at this point.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
I made a few comments, but I don't think there are any issues with this that need to be fixed so I merged it. Thanks @yyanyy!
